### PR TITLE
SYNPY-1103 multipart copy support

### PR DESCRIPTION
**Pull request by jkiang13**
Part 1 of 2 PRs for [SYNPY-1103](https://sagebionetworks.jira.com/browse/SYNPY-1103)

This PR refactors the existing multipart upload code to also support [multipart copies](https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/file/MultipartUploadCopyRequest.html).

### CODE REVIEW

File: synapseclient/core/upload/multipart_upload.py [hunk 1]
```diff
@@ -410,6 +415,18 @@ def _get_data_chunk(data, part_number, chunk_size):
...
 
 
+def _get_part_size(part_size, file_size):
+    part_size = part_size or DEFAULT_PART_SIZE
+
+    # can't exceed the maximum allowed num parts
+    part_size = max(
+        part_size,
+        MIN_PART_SIZE,
+        int(math.ceil(file_size / MAX_NUMBER_OF_PARTS))
```
Comment (zimingd) [inline synapseclient/core/upload/multipart_upload.py:425]: test changing default part size for copy to 100MB

File: synapseclient/core/upload/multipart_upload.py [hunk 2]
```diff
@@ -523,72 +561,119 @@ def multipart_upload_string(
...
 
+    part_size = _get_part_size(part_size, file_size)
+
+    upload_request = {
+        'concreteType': concrete_types.MULTIPART_UPLOAD_REQUEST,
+        'contentType': content_type,
+        'contentMD5Hex': md5_hex,
+        'fileName': dest_file_name,
+        'fileSizeBytes': file_size,
+        'generatePreview': preview,
```
Comment (zimingd) [inline synapseclient/core/upload/multipart_upload.py:572]: add `part_size` to `upload_request `

File: synapseclient/core/upload/multipart_upload.py [hunk 3]
```diff
@@ -523,72 +561,119 @@ def multipart_upload_string(
...
+        'partSizeBytes': part_size,
+        'sourceFileHandleAssociation': source_file_handle_association,
+        'storageLocationId': storage_location_id,
+    }
+
+    def md5_fn(_, response):
+        # for a multipart copy we use the md5 returned by the UploadPartCopy command
+        # when we add the part to the Synapse upload
+        md5_hex = re.search('(?<=<ETag>).*?(?=<\\/ETag>)', (response.content.decode('utf-8'))).group(0)
```
Comment (zimingd) [inline synapseclient/core/upload/multipart_upload.py:624]: explain regex in comment.

File: synapseclient/core/upload/multipart_upload.py [hunk 4]
```diff
@@ -523,72 +561,119 @@ def multipart_upload_string(
...
+        return md5_hex.replace('&quot;', '')
```
Comment (zimingd) [inline synapseclient/core/upload/multipart_upload.py:625]: also replace the unescaped double quote `"` character (just in case)

File: tests/unit/synapseclient/core/upload/unit_test_multipart_upload.py [hunk 5]
```diff
@@ -682,30 +744,127 @@ def test_multipart_upload_string(self):
...
+            expected_upload_request = {
+                'concreteType': 'org.sagebionetworks.repo.model.file.MultipartUploadCopyRequest',
+                'fileName': kwargs['dest_file_name'],
+                'generatePreview': kwargs['preview'],
+                'partSizeBytes': kwargs['part_size'],
+                'sourceFileHandleAssociation': source_file_handle_association,
+                'storageLocationId': kwargs['storage_location_id'],
+            }
+
+            multipart_copy(
```
Comment (zimingd) [inline tests/unit/synapseclient/core/upload/unit_test_multipart_upload.py:845]: split into own test. include case being tested either in name or as a comment.

File: tests/unit/synapseclient/core/upload/unit_test_multipart_upload.py [hunk 6]
```diff
@@ -734,90 +892,103 @@ def test_multipart_upload(self):
...
+
+                expected_upload_request,
+                mock.ANY,
+                mock.ANY,
+
                 max_threads=in_max_threads,
                 force_restart=in_force_restart,
             )
 
+            expected_upload_request = {
```
Comment (zimingd) [inline tests/unit/synapseclient/core/upload/unit_test_multipart_upload.py:955]: this is looks identical to a variable assignment above. is this necessary/intentional?

**Review by zimingd — CHANGES_REQUESTED** [review]

File: synapseclient/core/upload/multipart_upload.py [hunk 7]
```diff
@@ -83,26 +85,26 @@ class UploadAttempt:
...
-        part_size: int,
-        md5_hex: str,
-        content_type: str,
-        preview: bool,
-        storage_location_id: str,
+        dest_file_name,
+        part_size,
+
+        upload_request_payload,
+        part_fn,
```
[the remark this answers is not in the document]
Comment (jkiang13) [reply synapseclient/core/upload/multipart_upload.py]: part_request_body_provider_fn is required now, multipart copy passes in a no-op implementation.

File: synapseclient/core/upload/multipart_upload.py (hunk 1 above)
Comment (jkiang13) [reply synapseclient/core/upload/multipart_upload.py:419]: Added a test for this, although it is skipped in normal testing since it is slow-ish.

File: tests/integration/synapseclient/core/upload/test_multipart_upload.py [hunk 8]
```diff
@@ -187,7 +187,7 @@ def test_multipart_copy(syn, project, schedule_for_cleanup):
 
     fhid = entity['dataFileHandleId']
 
-#    source_file_handle = syn._get_file_handle_as_creator(fhid)
+    #    source_file_handle = syn._get_file_handle_as_creator(fhid)
```
Comment (jkiang13) [inline tests/integration/synapseclient/core/upload/test_multipart_upload.py:190]: Why is this here?

**Review by zimingd — APPROVED** [review]
